pipeat.go

```diff
@@ -247,7 +247,7 @@ func (r *PipeReaderAt) ReadAt(p []byte, off int64) (int, error) {
 func (r *PipeReaderAt) Read(p []byte) (int, error) {
 	trace("read", len(p))
 	n, err := r.ReadAt(p, r.f.readeroff)
-	if err == nil {
+	if n > 0 {
```

Comment:
I think I'd like these to be a bit more descriptive/self-documenting.  Calling out the specific case.  Something along the lines of...

```go
if err == nil || (err == io.EOF && n > 0) {
    ...
```
Maybe a switch would read better?
```go
switch {
case err == nil:
case err == io.EOF && n > 0:
default:
    return n, err
}
w.f.dataLock.Lock()
...
```
Not sure which I like better... please implement it how you think best. The main thing is that I like having things more explicit.
Thanks!

Reply:
I think if n> 0 we have to update the offset whatever the error is.

See for example: 

https://github.com/pkg/sftp/blob/master/client.go#L947
https://github.com/pkg/sftp/blob/master/client.go#L1423
https://github.com/pkg/sftp/blob/master/client.go#L1628

Reply:
I thought io.EOF was the only valid case where it could return `n > 0` with an error, but reviewing the docs/code that doesn't seem to be the case.  It isn't explicit about other error types but it does seem like any other case would be a catastrophic error that would probably prevent progress (hardware errors and such).

@drakkan .. Do you think simplifying the second test to just `n > 0` would be safe? It's been a while since I've worked with this code and you have more recent direct experience so I'll take you're word on the best semantics. Thanks.

Reply:
@eikenb the error we have seen here is indeed an `io.EOF` and this is the most common case. So advancing the offset after an `EOF` will fix the reported issue. 

I agree that other types of errors could only occur in case of hardware failure, disk full and the like. 

However, if we return n> 0 and an error, we also return some bytes to the reader, so I would always advance the offset if n> 0 (both reading and writing) as pkg/sftp does. pkg/sftp is much more used / tested than pipeat so I think we can trust it.

In this way the code is very protective and I think it is just a matter of preference, as stated above advancing the offset on `io.EOF` is enough to fix the reported issue.